Features/ManageabilityPkg/Universal/IpmiSmbiosTransferDxe/IpmiSmbiosTransferDxe.c

```diff
@@ -0,0 +1,293 @@
+
+  @param  Event    The event of notify protocol.
+  @param  Context  Notify event context.
+**/
+VOID
+EFIAPI
+IpmiSmbiosTransferSendTables (
+  IN EFI_EVENT  Event,
+  IN VOID       *Context
+  )
+{
+  EFI_STATUS                         Status;
+  SMBIOS_TABLE_3_0_ENTRY_POINT       *Smbios30Table;
+  SMBIOS_TABLE_3_0_ENTRY_POINT       *Smbios30TableModified;
+  EDKII_IPMI_BLOB_TRANSFER_PROTOCOL  *IpmiBlobTransfer;
+  UINT16                             Index;
+  UINT16                             SessionId;
+  UINT8                              *SendData;
+  UINT32                             SendDataSize;
+  UINT32                             RemainingDataSize;
+  BOOLEAN                            SmbiosTransferRequired;
+
+  gBS->CloseEvent (Event);
+
+  Status = gBS->LocateProtocol (&gEdkiiIpmiBlobTransferProtocolGuid, NULL, (VOID **)&IpmiBlobTransfer);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: No IpmiBlobTransferProtocol available. Exiting\n", __func__));
+    goto ErrorExit;
+  }
+
+  SmbiosTransferRequired = TRUE;
+  Smbios30Table          = NULL;
+  Status                 = EfiGetSystemConfigurationTable (&gEfiSmbios3TableGuid, (VOID **)&Smbios30Table);
+  if (EFI_ERROR (Status) || (Smbios30Table == NULL)) {
+    DEBUG ((DEBUG_ERROR, "%a: No SMBIOS Table found: %r\n", __func__, Status));
+    REPORT_STATUS_CODE_WITH_EXTENDED_DATA (
+      EFI_ERROR_CODE | EFI_ERROR_MAJOR,
+      EFI_SOFTWARE_EFI_APPLICATION,
+      SMBIOS_EC_DESC_NO_SMBIOS_TABLE,
+      sizeof (SMBIOS_EC_DESC_NO_SMBIOS_TABLE)
+      );
+    return;
+  }
+
+  //
+  // BMC expects the Smbios Entry Point to point to the address within the binary data sent
+  // The value is initially pointing to the location in memory where the table lives
+  // So we will save off that value and then modify the entry point to make the BMC happy
+  Smbios30TableModified = AllocateZeroPool (sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT));
+  CopyMem (Smbios30TableModified, Smbios30Table, sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT));
+  Smbios30TableModified->TableAddress = sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT);
+  //
+  // Fixup checksums in the Entry Point Structure
+  //
+  Smbios30TableModified->EntryPointStructureChecksum = 0;
+  Smbios30TableModified->EntryPointStructureChecksum =
+    CalculateCheckSum8 ((UINT8 *)Smbios30TableModified, Smbios30TableModified->EntryPointLength);
+
+  SendData = AllocateZeroPool (sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT) + Smbios30Table->TableMaximumSize);
+  CopyMem (SendData, Smbios30TableModified, sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT));
+  CopyMem (SendData + sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT), (UINT8 *)Smbios30Table->TableAddress, Smbios30Table->TableMaximumSize);
+  SendDataSize      = sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT) + Smbios30Table->TableMaximumSize;
+  RemainingDataSize = SendDataSize;
+
+  if (PcdGetBool (PcdSendSmbiosOnChanged)) {
+    SmbiosTransferRequired = DetectSmbiosChange (SendData, SendDataSize);

```

Comment:
Should the input be unmodified SMBIOS data? So, we can exit sooner if the transfer is not required.

Reply:
If we use unmodified SMBIOS data (variable Smbios30Table), we still need same process in DetectSmbiosChange(). We still have to put SMBIOS entry point buffer and data buffer into continuous buffer so we can calculate its hash value.